File: Algorithms/searching.c
/**
 * Searching Algorithms
 * Common searching algorithms for junior developer interviews
 */

#include <stdio.h>
#include <math.h>

/**
 * Linear/Sequential Search
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
int linearSearch(int arr[], int n, int target) {
    for (int i = 0; i < n; i++) {
        if (arr[i] == target) {
            return i;  // Return index if found
        }
    }
    return -1;  // Return -1 if not found
}

/**
 * Binary Search (Iterative)
 * Time Complexity: O(log n)
 * Space Complexity: O(1)
 * Note: Array must be sorted
 */
int binarySearchIterative(int arr[], int n, int target) {
    int left = 0;
    int right = n - 1;
    
    while (left <= right) {
        int mid = left + (right - left) / 2;
        
        if (arr[mid] == target) {
            return mid;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

/**
 * Binary Search (Recursive)
 * Time Complexity: O(log n)
 * Space Complexity: O(log n)
 * Note: Array must be sorted
 */
int binarySearchRecursiveHelper(int arr[], int target, int left, int right) {
    if (left > right) {
        return -1;
    }
    
    int mid = left + (right - left) / 2;
    
    if (arr[mid] == target) {
        return mid;
    } else if (arr[mid] < target) {
        return binarySearchRecursiveHelper(arr, target, mid + 1, right);
    } else {
        return binarySearchRecursiveHelper(arr, target, left, mid - 1);
    }
}

int binarySearchRecursive(int arr[], int n, int target) {
    return binarySearchRecursiveHelper(arr, target, 0, n - 1);
}

/**
 * Jump Search
 * Time Complexity: O(√n)
 * Space Complexity: O(1)
 * Note: Array must be sorted
 */
int jumpSearch(int arr[], int n, int target) {
    int step = (int)sqrt(n);
    int prev = 0;
    
    // Find the block where element is present
    while (arr[(step < n ? step : n) - 1] < target) {
        prev = step;
        step += (int)sqrt(n);
        if (prev >= n) {
            return -1;
        }
    }
    
    // Linear search in the identified block
    while (arr[prev] < target) {
        prev++;
        if (prev == (step < n ? step : n)) {
            return -1;
        }
    }
    
    if (arr[prev] == target) {
        return prev;
    }
    
    return -1;
}

/**
 * Interpolation Search
 * Time Complexity: O(log log n) for uniformly distributed data
 * Space Complexity: O(1)
 * Note: Array must be sorted
 */
int interpolationSearch(int arr[], int n, int target) {
    int left = 0;
    int right = n - 1;
    
    while (left <= right && target >= arr[left] && target <= arr[right]) {
        if (left == right) {
            if (arr[left] == target) {
                return left;
            }
            return -1;
        }
        
        // Probing the position
        int pos = left + ((target - arr[left]) * (right - left) / 
                        (arr[right] - arr[left]));
        
        if (arr[pos] == target) {
            return pos;
        } else if (arr[pos] < target) {
            left = pos + 1;
        } else {
            right = pos - 1;
        }
    }
    return -1;
}

// Helper function to print array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

// Test examples
int main() {
    printf("============================================================\n");
    printf("SEARCHING ALGORITHMS\n");
    printf("============================================================\n");
    
    int sortedArr[] = {2, 5, 8, 12, 16, 23, 38, 45, 56, 67, 78};
    int unsortedArr[] = {45, 23, 78, 12, 67, 5, 38, 16, 2, 56, 8};
    int n = 11;
    int target = 23;
    
    printf("\nSorted Array: ");
    printArray(sortedArr, n);
    printf("Unsorted Array: ");
    printArray(unsortedArr, n);
    printf("Target Element: %d\n\n", target);
    
    printf("1. Linear Search (works on unsorted array):\n");
    printf("------------------------------------------------------------\n");
    int result = linearSearch(unsortedArr, n, target);
    printf("   Found at index: %d\n", result);
    
    printf("\n2. Binary Search - Iterative (requires sorted array):\n");
    printf("------------------------------------------------------------\n");
    result = binarySearchIterative(sortedArr, n, target);
    printf("   Found at index: %d\n", result);
    
    printf("\n3. Binary Search - Recursive (requires sorted array):\n");
    printf("------------------------------------------------------------\n");
    result = binarySearchRecursive(sortedArr, n, target);
    printf("   Found at index: %d\n", result);
    
    printf("\n4. Jump Search (requires sorted array):\n");
    printf("------------------------------------------------------------\n");
    result = jumpSearch(sortedArr, n, target);
    printf("   Found at index: %d\n", result);
    
    printf("\n5. Interpolation Search (requires sorted array):\n");
    printf("------------------------------------------------------------\n");
    result = interpolationSearch(sortedArr, n, target);
    printf("   Found at index: %d\n", result);
    
    printf("\n============================================================\n");
    int targetNotFound = 100;
    printf("Searching for element not in array: %d\n", targetNotFound);
    printf("Linear Search: %d\n", linearSearch(sortedArr, n, targetNotFound));
    printf("Binary Search: %d\n", binarySearchIterative(sortedArr, n, targetNotFound));
    printf("============================================================\n");
    
    return 0;
}

File: Algorithms/searching.py
"""
Searching Algorithms
Common searching algorithms for junior developer interviews
"""


def linear_search(arr, target):
    """
    Linear/Sequential Search
    Time Complexity: O(n)
    Space Complexity: O(1)
    Description: Search for element by checking each element sequentially
    """
    for i in range(len(arr)):
        if arr[i] == target:
            return i  # Return index if found
    return -1  # Return -1 if not found


def binary_search_iterative(arr, target):
    """
    Binary Search (Iterative)
    Time Complexity: O(log n)
    Space Complexity: O(1)
    Description: Divide and conquer on sorted array
    Note: Array must be sorted
    """
    left = 0
    right = len(arr) - 1
    
    while left <= right:
        mid = left + (right - left) // 2
        
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    
    return -1


def binary_search_recursive(arr, target, left=None, right=None):
    """
    Binary Search (Recursive)
    Time Complexity: O(log n)
    Space Complexity: O(log n) - due to recursion stack
    Description: Recursive implementation of binary search
    Note: Array must be sorted
    """
    if left is None:
        left = 0
    if right is None:
        right = len(arr) - 1
    
    if left > right:
        return -1
    
    mid = left + (right - left) // 2
    
    if arr[mid] == target:
        return mid
    elif arr[mid] < target:
        return binary_search_recursive(arr, target, mid + 1, right)
    else:
        return binary_search_recursive(arr, target, left, mid - 1)


def jump_search(arr, target):
    """
    Jump Search
    Time Complexity: O(√n)
    Space Complexity: O(1)
    Description: Jump ahead by fixed steps, then linear search
    Note: Array must be sorted
    """
    import math
    
    n = len(arr)
    step = int(math.sqrt(n))
    prev = 0
    
    # Find the block where element is present
    while arr[min(step, n) - 1] < target:
        prev = step
        step += int(math.sqrt(n))
        if prev >= n:
            return -1
    
    # Linear search in the identified block
    while arr[prev] < target:
        prev += 1
        if prev == min(step, n):
            return -1
    
    if arr[prev] == target:
        return prev
    
    return -1


def interpolation_search(arr, target):
    """
    Interpolation Search
    Time Complexity: O(log log n) for uniformly distributed data
    Space Complexity: O(1)
    Description: Improved binary search for uniformly distributed sorted arrays
    Note: Array must be sorted
    """
    left = 0
    right = len(arr) - 1
    
    while left <= right and target >= arr[left] and target <= arr[right]:
        if left == right:
            if arr[left] == target:
                return left
            return -1
        
        # Probing the position with keeping uniform distribution in mind
        pos = left + ((target - arr[left]) * (right - left) // 
                     (arr[right] - arr[left]))
        
        if arr[pos] == target:
            return pos
        elif arr[pos] < target:
            left = pos + 1
        else:
            right = pos - 1
    
    return -1


# Test examples
if __name__ == "__main__":
    print("=" * 60)
    print("SEARCHING ALGORITHMS")
    print("=" * 60)
    
    # Test array (sorted for binary search)
    sorted_arr = [2, 5, 8, 12, 16, 23, 38, 45, 56, 67, 78]
    unsorted_arr = [45, 23, 78, 12, 67, 5, 38, 16, 2, 56, 8]
    
    target = 23
    
    print(f"\nSorted Array: {sorted_arr}")
    print(f"Unsorted Array: {unsorted_arr}")
    print(f"Target Element: {target}\n")
    
    print("1. Linear Search (works on unsorted array):")
    print("-" * 60)
    result = linear_search(unsorted_arr, target)
    print(f"   Found at index: {result}")
    
    print("\n2. Binary Search - Iterative (requires sorted array):")
    print("-" * 60)
    result = binary_search_iterative(sorted_arr, target)
    print(f"   Found at index: {result}")
    
    print("\n3. Binary Search - Recursive (requires sorted array):")
    print("-" * 60)
    result = binary_search_recursive(sorted_arr, target)
    print(f"   Found at index: {result}")
    
    print("\n4. Jump Search (requires sorted array):")
    print("-" * 60)
    result = jump_search(sorted_arr, target)
    print(f"   Found at index: {result}")
    
    print("\n5. Interpolation Search (requires sorted array):")
    print("-" * 60)
    result = interpolation_search(sorted_arr, target)
    print(f"   Found at index: {result}")
    
    # Test with element not in array
    print("\n" + "=" * 60)
    target_not_found = 100
    print(f"Searching for element not in array: {target_not_found}")
    print(f"Linear Search: {linear_search(sorted_arr, target_not_found)}")
    print(f"Binary Search: {binary_search_iterative(sorted_arr, target_not_found)}")
    print("=" * 60)
